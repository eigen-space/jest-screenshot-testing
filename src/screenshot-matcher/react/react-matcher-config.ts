import { ComponentClass } from 'react';
import { Config } from '../../common/config/config';

export class ReactMatcherConfig {
    static globalStyles: any;
    static serializer: any;
    static themeWrapper: ComponentClass;

    // noinspection JSUnusedGlobalSymbols
    static set screenshoterUrl(value: string) {
        Config.screenshoterUrl = value;
    }
}