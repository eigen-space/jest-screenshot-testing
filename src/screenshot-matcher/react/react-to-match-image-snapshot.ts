import { cloneElement, createElement, ReactElement } from 'react';
import { CommonWrapper, mount, ReactWrapper, ShallowWrapper } from 'enzyme';
import { create, ReactTestRendererJSON } from 'react-test-renderer';
import toJson from 'enzyme-to-json';
import { MatcherState } from 'jest-types-workaround';
import { Html } from '../../@types/common';
import { Dictionary } from '../../common/types/dictionary';
import { matchScreenshot } from '../screenshot-matcher';
import { renderToStaticMarkup } from 'react-dom/server';
import { ReactMatcherConfig } from './react-matcher-config';

interface ToMatchArgs {
    component: ReactElement<Dictionary> | ReactWrapper | ShallowWrapper;
    commonStyles: CSS;
    props: Dictionary;
    componentSource?: ReactElement<Dictionary>;
}

export async function toMatchComponentImageAsyncReact(this: MatcherState, data: ToMatchArgs): Promise<Dictionary> {
    if (!ReactMatcherConfig.serializer) {
        throw new Error('Serializer styles is not set. Please, set it in ReactMatcherConfig.');
    }

    const { component, commonStyles, props, componentSource } = data;

    let preparedComponent = component;
    let rawHtml: Html;
    let json: ReactTestRendererJSON | null;

    if (component instanceof ReactWrapper) {
        rawHtml = component.html();
        json = toJson(component);
    } else if (component instanceof ShallowWrapper && componentSource) {
        const mountComponent = mount<Dictionary>(componentSource);
        mountComponent.setProps(props);
        mountComponent.instance().forceUpdate();
        const state = component.instance().state;
        mountComponent.instance().setState(state);
        rawHtml = mountComponent.html();
        json = toJson(mountComponent as CommonWrapper);
    } else {
        const componentClone = cloneElement(component as ReactElement<Dictionary>, props);

        if (ReactMatcherConfig.themeWrapper) {
            preparedComponent = createElement(ReactMatcherConfig.themeWrapper, {}, componentClone);
        } else {
            preparedComponent = componentClone;
        }
        rawHtml = renderToStaticMarkup(preparedComponent);
        json = create(preparedComponent).toJSON();
    }

    const styleSnapshot = ReactMatcherConfig.serializer.print(json, () => rawHtml);
    const separationIndex = styleSnapshot.indexOf('<');
    const html = styleSnapshot.slice(separationIndex);
    const css = styleSnapshot.slice(0, separationIndex);

    let globalStyles = '';
    if (ReactMatcherConfig.globalStyles) {
        // There no correct interface for GlobalStyles.
        // tslint:disable-next-line:no-any
        globalStyles += (ReactMatcherConfig.globalStyles as any).rules.join('');
    }

    if (commonStyles) {
        globalStyles += commonStyles;
    }

    if (globalStyles) {
        globalStyles += '\n';
    }

    return matchScreenshot(this, html, `${globalStyles}${css}`);
}