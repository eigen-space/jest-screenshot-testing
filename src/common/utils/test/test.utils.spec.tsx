import * as React from 'react';
import { TestUtils } from './test.utils';
import * as renderer from 'react-test-renderer';

describe('TestUtils', () => {

    describe('#boxWrap', () => {
        const component = <div>Component</div>;

        it('should set defined box value', async () => {
            const box = { width: '43', height: '22' };
            const json = renderer.create(TestUtils.boxWrap(component, box)).toJSON();
            await expect({ component: json }).toMatchSnapshot();
        });

        it('should set default value, any key in box undefined', async () => {
            const box = {};
            const json = renderer.create(TestUtils.boxWrap(component, box)).toJSON();
            await expect({ component: json }).toMatchSnapshot();
        });
    });
});
