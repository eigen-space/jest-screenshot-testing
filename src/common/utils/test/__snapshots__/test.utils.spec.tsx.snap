// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestUtils #boxWrap should set default value, any key in box undefined 1`] = `
Object {
  "component": <div
    style={
      Object {
        "height": "auto",
        "width": "fit-content",
      }
    }
  >
    <div>
      Component
    </div>
  </div>,
}
`;

exports[`TestUtils #boxWrap should set defined box value 1`] = `
Object {
  "component": <div
    style={
      Object {
        "height": "22px",
        "width": "43px",
      }
    }
  >
    <div>
      Component
    </div>
  </div>,
}
`;
